File: 200-num-Islands/solution.js
const directions = [[0, -1], [0, 1], [-1, 0], [1, 0]]
var rows
var cols
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    rows = grid.length
    cols = grid[0].length
    let color = Array(rows).fill(null).map(() => Array(cols).fill(0))

    let result = 0
    for (let i = 0; i < rows; i++)
        for (let j = 0; j < cols; j++)
            if (grid[i][j] === '1' && color[i][j] === 0){
                bfs(grid, color, i, j)
                result++
            }
    return result
}
var bfs = function (grid, color, i, j){
    color[i][j] = 1
    let queue = []
    queue.push([i, j])
    while (queue.length > 0){
        let layerSize = queue.length
        for (let k = 0; k < layerSize; k++) {
            let cur = queue.shift()
            let curX = cur[0]
            let curY = cur[1]
            for (let direction of directions){
                let newX = curX + direction[0]
                let newY = curY + direction[1]
                if (inGrid(newX, newY) && grid[newX][newY] === '1' && color[newX][newY] === 0){
                    queue.push([newX, newY])
                    color[newX][newY] = 1
                }
            }
        }
    }
}

const inGrid = (i, j) => i >= 0 && j >= 0 && i < rows && j < cols
if(require.main === module){
    let grid = [
        ['1','1','1','1','0'],
        ['1','1','0','1','0'],
        ['1','1','0','0','0'],
        ['0','0','0','0','0']
    ]
    console.log("Ans = ", numIslands(grid))
}
/*
自己寫的
相似題:
695-max-area-of-island/solution.go
 */

File: 200-num-Islands/solution.go
package main

import "fmt"

var directions [][]int = [][]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}
var rows int
var cols int

func numIslands(grid [][]byte) int {
	count := 0
	rows = len(grid)
	cols = len(grid[0])
	inGrid := func(i int, j int) bool {
		return i >= 0 && j >= 0 && i < rows && j < cols
	}
	var dfs func(i int, j int)
	dfs = func(i int, j int) {
		grid[i][j] = '0'
		for _, direction := range directions {
			newX := i + direction[0]
			newY := j + direction[1]
			if inGrid(newX, newY) && grid[newX][newY] == '1' {
				dfs(newX, newY)
			}
		}
	}
	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			if grid[i][j] == '1' {
				dfs(i, j)
				count++
			}
		}
	}
	return count
}
func main() {
	grid := [][]byte{
		{'1', '1', '1', '1', '0'},
		{'1', '1', '0', '1', '0'},
		{'1', '1', '0', '0', '0'},
		{'0', '0', '0', '1', '0'},
	}
	fmt.Println("Ans = ", numIslands(grid))
}

/*
自己寫的

走過的地方就不應該再走, 如果是'1'的話, 直接改成'0'
從當前座標出發, 遍歷四個方向的所有座標, 在grid範圍內值為'1'的座標才值得被遍歷
如果一個座標的上下左右都被遍歷過代表這個島嶼已經被我們用dfs走完了, 就可以返回了

bfs看js
dfs看go
*/
